assertj-core/src/main/java/org/assertj/core/api/recursive/comparison/ComparingNormalizedFields.java

```diff
@@ -0,0 +1,35 @@
+package org.assertj.core.api.recursive.comparison;
+
+import static java.util.stream.Collectors.toSet;
+import static org.assertj.core.util.introspection.PropertyOrFieldSupport.COMPARISON_NORMALIZED;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.assertj.core.internal.Objects;
+
+/**
+ * A {@link RecursiveComparisonIntrospectionStrategy} that introspects fields provided their normalized name.
+ */
+abstract class ComparingNormalizedFields implements RecursiveComparisonIntrospectionStrategy {
+
+  @Override
+  public Set<String> getChildrenNodeNamesOf(Object node) {
+    if (node == null) return new HashSet<>();
+    Set<String> fieldsNames = Objects.getFieldsNames(node.getClass());
+    return fieldsNames.stream().map(this::normalizeFieldName).collect(toSet());
+  }
+
+  protected abstract String normalizeFieldName(String fieldName);
+
+  @Override
+  public Object getChildNodeValue(String childNodeName, Object instance) {
+    return COMPARISON_NORMALIZED.getSimpleValue(childNodeName, instance);
```

Comment:
And it is consistent as the `normalizeFieldName` method is already being used within `CaseStyleInsensitiveFieldLookupStrategy`, which is the `FieldLookUpStrategy` for `FieldSupport.COMPARISON_NORMALIZED`. I think that the abstraction you're looking for is indeed `FieldLookUpStrategy`. Thanks to it you can find any field inside a class regardless of the case format. Otherwise (with the try/catch approach) you'd have to generate all the possible/probable variations of the same name: `lastName`, `last_name` (these first two would be the cases addresses by the try/catch) but also the less frequent but not so unlikely `_last_name`, `last_name_` or even `_last_name_`

Reply:
That's not entirely correct, what you are saying is that the abstract `normalizeFieldName` is consistent with `CaseStyleInsensitiveFieldLookupStrategy` which uses `ComparingSnakeOrCamelCaseFields.normalizeFieldName` so at this point this does not honor a different implementation of `normalizeFieldName`.

Take this example where I implemented a simple `normalizeFieldName` that simply lowercase fields name before comparison

```java
  @Test
  void should_pass_with_lowercase_compared_fields() {
    // GIVEN
    Animal fox = new Animal("fox");
    AnimalDto foxDto = new AnimalDto("fox");
    // WHEN/THEN
    then(foxDto).usingRecursiveComparison()
             .withIntrospectionStrategy(new ComparingLowercaseNormalizedFields())
             .isEqualTo(fox);
  }

  static class Animal {
    String raceName;

    Animal(String raceName) {
      this.raceName = raceName;
    }
  }

  static class AnimalDto {
    String racename;

    AnimalDto(String racename) {
      this.racename = racename;
    }
  }
```
the test fails with:
```
org.assertj.core.util.introspection.IntrospectionError: 
Can't find any field or property with name 'racename'.
Error when introspecting properties was :
- No getter for property 'racename' in org.assertj.core.api.recursive.comparison.RecursiveComparisonAssert_isEqualTo_withIntrospectionStrategy_Test$Animal 
Error when introspecting fields was :
- Unable to obtain the value of the field <'racename'> from <org.assertj.core.api.recursive.comparison.RecursiveComparisonAssert_isEqualTo_withIntrospectionStrategy_Test$Animal@64dafeed>
```
If the lowercase normalization had been honored, `Animal.raceName` would have been found when searching for `racename`

Reply:
I think to support a generic `ComparingNormalizedFields` we also need a `revertNormalize` method with a `getChildNodeValue` implemented as 
```java
  public Object getChildNodeValue(String childNodeName, Object instance) {
    try {
      return COMPARISON.getSimpleValue(childNodeName, instance);
    } catch (Exception e) {
      return COMPARISON.getSimpleValue(revertNormalize(childNodeName), instance);
    }
  }
```
The reason is in my example, we try to find `racename` field in `Animal` which has a `raceName` field, by using `revertNormalize("racename")` we would get `raceName` and find the field value.

Now in my `ComparingLowercaseNormalizedFields` implementation, there is no easy way to implement `revertNormalize`, whereas normalizing snake case to camel case can be reverted to simply camel case to snake case.

Reply:
I got it to work with adding a `protected Collection<String> revertNormalizeFieldName(String fieldName)` to `ComparingNormalizedFields`, I had to return a list in case several different fields led to the same normalization that would be used in `ComparingNormalizedFields.getChildNodeValue` implementation:

```java
  public Object getChildNodeValue(String childNodeName, Object instance) {
    try {
      return COMPARISON.getSimpleValue(childNodeName, instance);
    } catch (Exception e) {

      Collection<String> originalFields = revertNormalizeFieldName(childNodeName);
      for (String originalField : originalFields) {
        try {
          return COMPARISON.getSimpleValue(originalField, instance);
        } catch (Exception ex) {
          // ignore to test the next field
        }
      }
    }
    throw new IntrospectionError("Unable to find " + childNodeName + " field in " + instance);
  }
```


```java
  static class ComparingLowercaseNormalizedFields extends ComparingNormalizedFields {

    MultiValuedMap<String, String> originalFieldNamesByNormalizedFieldName =  MultiMapUtils.newListValuedHashMap();

    @Override
    public String normalizeFieldName(String name) {
      String lowerCase = name.toLowerCase();
      originalFieldNamesByNormalizedFieldName.put(lowerCase, name);
      return lowerCase;
    }

    @Override
    protected Collection<String> revertNormalizeFieldName(String fieldName) {
      return originalFieldNamesByNormalizedFieldName.get(fieldName);
    }
  }
```
In my example first implementation I started with `String revertNormalizeFieldName(String)` implemented with a simple `Map<String, String>` map but that did not work, the first normalized field was `Animal.raceName` so the map had the entry `racename` to `raceName` which is fine until the `AnimalDto.raceName` was normalized, it replaced the entry `racename` to `raceName` by `racename` to `racename` and when it came to find the field `racename` in `Animal.raceName` that could not work.

Now I feel it might a bit too complicated for users, thoughts?